### PR TITLE
Variables: Extend option pickers to accept custom onChange callback

### PR DESCRIPTION
**Pull request by Clarity-89**
**What this PR does / why we need it**:
Add optional `onVariableChange` prop to option pickers. This will enable to control the setting of the variable and do some extra post-processing if necessary. 

Also some minor changes to DashboardPicker to make it work better inside forms. 

**Which issue(s) this PR fixes**:

### CODE REVIEW

File: public/app/features/variables/pickers/OptionsPicker/actions.ts [hunk 1]
```diff
@@ -72,7 +72,13 @@ export const filterOrSearchOptions = (searchQuery = ''): ThunkResult<void> => {
   };
 };
 
-export const commitChangesToVariable = (): ThunkResult<void> => {
+const setVariable = async (updated: any) => {
```
Comment (hugohaggmark) [inline public/app/features/variables/pickers/OptionsPicker/actions.ts:75]: This should be VariableWithMultiSupport, right?
```suggestion
const setVariable = async (updated: VariableWithMultiSupport) => {
```

File: public/app/features/variables/pickers/OptionsPicker/actions.ts [hunk 2]
```diff
@@ -72,7 +72,13 @@ export const filterOrSearchOptions = (searchQuery = ''): ThunkResult<void> => {
...
+  const adapter = variableAdapters.get(updated.type);
+  await adapter.setValue(updated, updated.current, true);
+  return;
+};
+
+export const commitChangesToVariable = (callback?: (updated: any) => void): ThunkResult<void> => {
```
Comment (hugohaggmark) [inline public/app/features/variables/pickers/OptionsPicker/actions.ts:81]: Same here:
```suggestion
export const commitChangesToVariable = (callback?: (updated: VariableWithMultiSupport) => void): ThunkResult<void> => {
```

**Review by hugohaggmark — COMMENTED** [review]
Looks ok so far and I'm going to test this now. I'll get back with more feedback after that.

File: public/app/features/variables/pickers/OptionsPicker/actions.ts (hunk 1 above)
Comment (Clarity-89) [reply public/app/features/variables/pickers/OptionsPicker/actions.ts:75]: Great catch! Forgot to update the types.

File: public/app/features/variables/pickers/OptionsPicker/actions.ts [hunk 3]
```diff
@@ -88,9 +94,11 @@ export const commitChangesToVariable = (): ThunkResult<void> => {
...
     }
 
-    const adapter = variableAdapters.get(updated.type);
-    await adapter.setValue(updated, updated.current, true);
-    return;
+    if (callback) {
+      return callback(updated);
+    } else {
+      return setVariable(updated);
+    }
```
Comment (hugohaggmark) [inline public/app/features/variables/pickers/OptionsPicker/actions.ts:101]: Nit: maybe we could rewrite this as:
```suggestion
    if (callback) {
      return callback(updated);
    } 
    
    return setVariable(updated);
```

File: public/app/features/variables/textbox/TextBoxVariablePicker.tsx [hunk 4]
```diff
@@ -27,7 +27,15 @@ export function TextBoxVariablePicker({ variable }: Props): ReactElement {
...
-    variableAdapters.get(variable.type).updateOptions(variable);
+
+    if (onVariableChange) {
+      onVariableChange({
+        ...variable,
+        current: { ...variable.current, value: updatedValue },
+      });
+    } else {
+      variableAdapters.get(variable.type).updateOptions(variable);
+    }
```
Comment (hugohaggmark) [inline public/app/features/variables/textbox/TextBoxVariablePicker.tsx:38]: Nit: maybe this is clearer:
```suggestion
    if (onVariableChange) {
      onVariableChange({
        ...variable,
        current: { ...variable.current, value: updatedValue },
      });
      return;
    } 
    
    variableAdapters.get(variable.type).updateOptions(variable);
```

**Review by hugohaggmark — APPROVED** [review]
LGTM! 🎉

**Review by torkelo — APPROVED** [review]

**Pull request merged**